lib/synapse/haproxy.rb

```diff
@@ -655,18 +693,33 @@ def generate_frontend_stanza(watcher, config)
     end
 
     def generate_backend_stanza(watcher, config)
+      backends = {}
+
+      # The ordering here is important.  First we add all the backends in the
+      # disabled state...
+      @seen.fetch(watcher.name, []).each do |backend_name, backend|
+        backends[backend_name] = backend.merge('enabled' => false)
+      end
+
+      # ... and then we overwite any backends that the watchers know about,
+      # setting the enabled state.
+      watcher.backends.each do |backend|
+        backend_name = construct_name(backend)
+        backends[backend_name] = backend.merge('enabled' => true)
+      end
+
       if watcher.backends.empty?
-        log.warn "synapse: no backends found for watcher #{watcher.name}"
+        log.debug "synapse: no backends found for watcher #{watcher.name}"
       end
 
       stanza = [
         "\nbackend #{watcher.name}",
         config.map {|c| "\t#{c}"},
-        watcher.backends.shuffle.map {|backend|
-          backend_name = construct_name(backend)
+        backends.map {|backend_name, backend|
```

Comment:
This is true if all HAProxy instances start up at exactly the same moment and request loads are such that you notice this. In our experience things even out on their own (both haproxy reloads take time and request loads to servers has jitter). We much more often run into situations where HAProxy is restarting a lot and the HAProxy stats page becomes really hard to read which can impede debugging efforts.

This is purely so that the stats page across reloads doesn't change a ton. I can change it back if you prefer the random behavior :-)